Added Crypto API reference documentation (#3952)
dapr/docs: daprdocs/content/en/reference/api/cryptography_api.md

---
type: docs
title: "Cryptography API reference"
linkTitle: "Cryptography API"
description: "Detailed documentation on the cryptography API"
weight: 1300
---

> Dapr provides cross-platform and cross-language support for encryption and decryption support via the 

Reviewer: Please add a note that the HTTP APIs are not recommended besides development and testing. Using the SDKs, which use the gRPC APIs, is strongly recommended

Author: As this is already merged, added this suggestion to https://github.com/dapr/docs/pull/3967

> Dapr provides native, cross-platform, and cross-language support for encryption and decryption support via the 

Reviewer 2: ```suggestion
Dapr provides cross-platform and cross-language support for encryption and decryption support via the 
```

> Cryptography building block. Besides the [language specific SDKs]({{<ref sdks>}}), a developer can invoke these capabilities using

Reviewer 2: ```suggestion
cryptography building block. Besides the [language specific SDKs]({{<ref sdks>}}), a developer can invoke these capabilities using
```

> the gRPC API endpoints below.

Reviewer 2: Do you mean to say HTTP rather than gRPC, Yes?
```suggestion
the HTTP API endpoints below.
```

| dapr-decryption-key-name      | If `dapr-omit-decryption-key-name` is true, this contains the name of the intended decryption key to include in the output.                                                                         |                                                                                   | Required only if `dapr-omit-decryption-key-name` is true |
| dapr-data-encryption-cipher   | The cipher to use for the encryption operation                                                                                                                                                      | `aes-gcm` or `chacha20-poly1305`                                                  | No                                                       |

### HTTP Response

#### Response Body
The response to an encryption request will have its content type header set to `application/octet-stream` as it

> will return an array of bytes with the encrypted payload.

Reviewer 2: ```suggestion
returns an array of bytes with the encrypted payload.
```

#### Response Codes
| Code | Description                                                             |
|------|-------------------------------------------------------------------------|
| 200  | OK                                                                      |
| 400  | Crypto provider not found                                               |

> | 500  | Request formatted correctly, error in dapr code or underlying component |

Reviewer 2: ```suggestion
| 500  | Request formatted correctly, error in Dapr code or underlying component |
```

```

## Decrypt Payload

This endpoint lets you decrypt a value provided as a byte array using a specified key and crypto component.

#### HTTP Request

```
PUT curl http://localhost:3500/v1.0/crypto/<crypto-store-name>/decrypt
```

#### URL Parameters

| Parameter         | Description                                                 |
|-------------------|-------------------------------------------------------------|
| daprPort          | The Dapr port                                               |
| crypto-store-name | The name of the crypto store to get the decryption key from |

> Note all parameters are case-sensitive.

#### Headers
Additional decryption parameters are configured by setting headers with the appropriate values. The following table
details the required and optional headers to set with every decryption request.

| Header Key    | Description                                              | Required |
|---------------|----------------------------------------------------------|----------|
| dapr-key-name | The name of the key to use for the decryption operation. | Yes      |

### HTTP Response

#### Response Body

> The response to a decryption request will have its content type header to set `application/octet-stream` as it will 

Reviewer 2: ```suggestion
The response to a decryption request will have its content type header to set `application/octet-stream` as it 
```

Author: Rejecting this suggestion as the next would needed to be updated from "return" to "returns" for the sentence to flow. Updated both it and your suggestion in a recent commit.

> return an array of bytes representing the decrypted payload.

Reviewer 2: ```suggestion
returns an array of bytes representing the decrypted payload.
```
